### DISCUSSION TITLE
Source: Input Manipulation

### DISCUSSION THREAD

**Discussion by PanicKk** [savonet/liquidsoap#3390] [User Support]
Is there a way to change the source of an input.rtmp even though it is already connected to an output?

I thought of having two input.rtmps and switch between those. The switch works fine and all, but after some time, I may want to change the source of the first input and then switch to that ( same for the second one ) while always keeping a connection to the output server without interruption between source switch.

Demonstration:
```
# The source of each input is an RTMP Server (Nginx with the RTMP module)
source_one = input.rtmp("rtmp://server:1935/live/key", listen=false)
source_two = input.rtmp("rtmp://server:1935/live/different-key", listen=false)

stream = switch(track_sensitive=false, [(booean1, source_one), (boolean2, source_two)]

# So after some time I decide to switch to source_two for a while, and later switch to source_one but with a different stream url
source_one = input.rtmp("rtmp://server:1935/live/new-key", listen=false)

# ... switch back and worth, with the same / different inputs, as desired
```
I tried using references:
```
input1 = ref("rtmp://server:1935/live/key)

input1 := "rtmp://server:1935/live/new-key"
```
But, even though the url changes, the input source is still the same, even after the switch.

Does it save the url in a cache / buffer once the output starts streaming? Is there a way to do that sort of switch / manipulation?

Maybe shutdown the first input, change the url, then start again?

I would really appreciate your help. If that is possible ofc...

**Comment by TheRealMattLear** [comment]
I think delay() with a fallback() might work for you as the example at https://www.liquidsoap.info/doc-1.4.4/cookbook.html#force-a-fileplaylist-to-be-played-at-least-every-xx-minutes

Not sure if this will work exactly but this is where i would start:

```
source_one = input.rtmp("rtmp://server:1935/live/key", listen=false)
source_two = input.rtmp("rtmp://server:1935/live/different-key", listen=false)

main_source = fallback([
    delay(60.,source_one),
    source_two
])
```

**Comment by vitoyucepi** [comment]
Hi @PanicKk,
I think the provided description could be satisfied by the [`source.dynamic`](https://www.liquidsoap.info/doc-2.2.0/reference.html#source.dynamic) function.
The dynamic sources are described on [page 207 of the book](https://www.liquidsoap.info/book/book.pdf).

I have designed a script that demonstrates how to dynamically change sources, which will be helpful for you.
```ruby
interactive.harbor(port=8000)
def dynamic_rtmp_source()
    rtmp_address = interactive.string("rtmp", "rtmp://0.0.0.0:8002/a/b")
    current_rtmp_address = ref(rtmp_address())
    current_rtmp_source = ref(input.rtmp(rtmp_address()))

    def update_rtmp_source()
        if rtmp_address() != current_rtmp_address() then
            current_rtmp_address := rtmp_address()
            current_rtmp_source().shutdown()
            current_rtmp_source := input.rtmp(rtmp_address())
        end
        current_rtmp_source()
    end
    source.dynamic(update_rtmp_source)
end

s = dynamic_rtmp_source()
output.harbor(%mp3, s, mount='/test', port=8001, fallible=true)

```

The idea is to create a function that will internally track the state of the `input.rtmp` source based on the `interactive.string`.
The `dynamic_rtmp_source` function works like this
1. Define `interactive.string` that contains the rtmp address for the `input.rtmp`.
1. Save the initial state of the interactive variable and construct the default rtmp source.
   I can define the rtmp source as `ref(null())`, but it will require additional null checks.
1. Define a function that will check if the state changed.
1. Return dynamic source that will periodically check if it'll require to change the internal source.

The `update_rtmp_source` function uses external vars as a state storage because it's the only way to preserve data changes. It works like this
1. Check if the new address set by `interactive.string` differs from the previous one.
   1. Store new address.
   1. Shut down the old source.
   1. Construct a new `input.rtmp` source.
1. Return the current value of the source.
   It can be the same as before if nothing changes.
   But it can also switch to a new one.


<details>
<summary>Are there any further enhancements?</summary>

1. Add some arguments to `dynamic_rtmp_source`. For example, the default rtmp address and interactive variable name.
1. Postpone old source shutdown by using `thread.run`.
1. Use `thread.when` to update the dynamic source using the `set` method.

```ruby
def dynamic_rtmp_source(name, default_rtmp_address)
    rtmp_address = interactive.string(name, default_rtmp_address)
    rtmp_source = ref(input.rtmp(rtmp_address()))
    dynamic_source = source.dynamic()
    dynamic_source.set(rtmp_source())

    def source_transition(old_source, new_source)
        dynamic_source.set(new_source)
        old_source.shutdown()
    end

    def on_address_changed()
        old_source = rtmp_source()
        new_source = input.rtmp(rtmp_address())
        rtmp_source := new_source
        thread.when({new_source.is_ready()}, {source_transition(old_source, new_source)}, once=true)
    end
    thread.when(getter.changes(rtmp_address), on_address_changed)

    dynamic_source
end
   
s = dynamic_rtmp_source("rtmp", "rtmp://0.0.0.0:8002/a/b")
output.harbor(%mp3, s, mount='/test', port=8001, fallible=true)
interactive.harbor(port=8000)
```

I don't know how to make a smooth transition between sources, but this one offers the best experience without interrupting the fallible source.
</details>
